File: Predictive_Demand_Model/demand_forecast.R
# =================================================================
# Nova Leisure Group — Predictive Demand Model (R)
# Author: Amogh H.H. — Commercial Finance / Business Analytics
#
# Purpose
#   Ingest Daily_Site_Footfall.csv and produce a 90-day site-level
#   demand forecast that feeds back into the Excel revenue line.
#
# Models
#   1. Holt-Winters triple exponential smoothing (additive, weekly
#      seasonality of 7).
#   2. ARIMA via auto.arima — model is auto-selected per site
#      based on AICc.
#   3. Naive Seasonal benchmark (sNaive) — used as accuracy baseline.
#
# Outputs
#   - forecast_output.csv   : site × date × point + 80/95% PIs
#   - model_metrics.txt     : per-model RMSE, MAE, MAPE on holdout
#   - residual_diagnostics  : Ljung-Box p-values (printed)
# =================================================================

suppressPackageStartupMessages({
  pkgs <- c("forecast", "dplyr", "lubridate", "readr", "tidyr")
  to_install <- setdiff(pkgs, rownames(installed.packages()))
  if (length(to_install)) install.packages(to_install, repos = "https://cloud.r-project.org")
  lapply(pkgs, library, character.only = TRUE)
})

# --- Config -------------------------------------------------------
HORIZON_DAYS  <- 90
HOLDOUT_DAYS  <- 28
SEED          <- 42
set.seed(SEED)

ROOT     <- dirname(dirname(normalizePath(sys.frame(1)$ofile, mustWork = FALSE)))
if (!nzchar(ROOT) || ROOT == "/") {
  ROOT <- "."
}
INPUT    <- file.path(ROOT, "Daily_Site_Footfall.csv")
OUTDIR   <- file.path(ROOT, "Predictive_Demand_Model")
if (!dir.exists(OUTDIR)) dir.create(OUTDIR, recursive = TRUE)

OUT_CSV  <- file.path(OUTDIR, "forecast_output.csv")
OUT_METR <- file.path(OUTDIR, "model_metrics.txt")
OUT_PLOT <- file.path(OUTDIR, "forecast_plot.pdf")

# --- Load ---------------------------------------------------------
df <- read_csv(INPUT, show_col_types = FALSE) %>%
  mutate(Date = as.Date(Date))

cat(sprintf("Loaded %d daily rows across %d sites.\n",
            nrow(df), length(unique(df$Site_Code))))

sites <- unique(df$Site_Code)

# --- Helpers ------------------------------------------------------
accuracy_metrics <- function(actual, pred) {
  err  <- actual - pred
  rmse <- sqrt(mean(err^2, na.rm = TRUE))
  mae  <- mean(abs(err), na.rm = TRUE)
  mape <- mean(abs(err / actual) * 100, na.rm = TRUE)
  c(RMSE = rmse, MAE = mae, MAPE = mape)
}

all_forecasts <- list()
metrics_rows  <- list()
diag_rows     <- list()

for (s in sites) {
  site_df <- df %>%
    filter(Site_Code == s) %>%
    arrange(Date)

  site_name <- unique(site_df$Site_Name)[1]
  region    <- unique(site_df$Region)[1]

  y <- ts(site_df$Visitors, frequency = 7)

  # Train / holdout
  n <- length(y)
  train <- window(y, end = c(floor((n - HOLDOUT_DAYS) / 7), ((n - HOLDOUT_DAYS) %% 7) + 1))
  # Simpler: by index
  train <- ts(head(site_df$Visitors, n - HOLDOUT_DAYS), frequency = 7)
  test  <- tail(site_df$Visitors, HOLDOUT_DAYS)

  # --- Model 1: Holt-Winters (additive, with damped trend off)
  hw_fit  <- HoltWinters(train, seasonal = "additive")
  hw_pred <- forecast(hw_fit, h = HOLDOUT_DAYS)$mean

  # --- Model 2: ARIMA (auto)
  ar_fit  <- auto.arima(train, seasonal = TRUE, stepwise = TRUE, approximation = FALSE)
  ar_pred <- forecast(ar_fit, h = HOLDOUT_DAYS)$mean

  # --- Model 3: Seasonal Naive benchmark
  sn_pred <- snaive(train, h = HOLDOUT_DAYS)$mean

  # --- Accuracy
  hw_m <- accuracy_metrics(test, hw_pred)
  ar_m <- accuracy_metrics(test, ar_pred)
  sn_m <- accuracy_metrics(test, sn_pred)

  metrics_rows[[length(metrics_rows) + 1]] <- data.frame(
    Site_Code = s, Site_Name = site_name,
    Model = c("HoltWinters", "ARIMA", "SeasonalNaive"),
    RMSE  = c(hw_m["RMSE"], ar_m["RMSE"], sn_m["RMSE"]),
    MAE   = c(hw_m["MAE"],  ar_m["MAE"],  sn_m["MAE"]),
    MAPE  = c(hw_m["MAPE"], ar_m["MAPE"], sn_m["MAPE"])
  )

  # --- Pick best model by MAPE
  best_idx <- which.min(c(hw_m["MAPE"], ar_m["MAPE"], sn_m["MAPE"]))
  best     <- c("HoltWinters", "ARIMA", "SeasonalNaive")[best_idx]

  # --- Ljung-Box on residuals
  resid <- switch(best,
                  HoltWinters  = residuals(hw_fit),
                  ARIMA        = residuals(ar_fit),
                  SeasonalNaive= rep(NA, length(train)))
  lb_p <- tryCatch(Box.test(resid, lag = 14, type = "Ljung-Box")$p.value,
                   error = function(e) NA)

  diag_rows[[length(diag_rows) + 1]] <- data.frame(
    Site_Code = s, Best_Model = best, LjungBox_p = lb_p
  )

  # --- Refit on FULL history and forecast forward 90 days
  y_full <- ts(site_df$Visitors, frequency = 7)
  final_fit <- switch(best,
                      HoltWinters = HoltWinters(y_full, seasonal = "additive"),
                      ARIMA       = auto.arima(y_full, seasonal = TRUE),
                      SeasonalNaive = NULL)
  if (best == "SeasonalNaive") {
    fc <- snaive(y_full, h = HORIZON_DAYS)
  } else {
    fc <- forecast(final_fit, h = HORIZON_DAYS, level = c(80, 95))
  }

  last_date <- max(site_df$Date)
  out_dates <- seq(last_date + 1, by = "day", length.out = HORIZON_DAYS)

  all_forecasts[[length(all_forecasts) + 1]] <- data.frame(
    Date          = out_dates,
    Site_Code     = s,
    Site_Name     = site_name,
    Region        = region,
    Model         = best,
    Forecast      = round(as.numeric(fc$mean)),
    Lower_80      = round(as.numeric(if (!is.null(fc$lower)) fc$lower[, 1] else fc$mean)),
    Upper_80      = round(as.numeric(if (!is.null(fc$upper)) fc$upper[, 1] else fc$mean)),
    Lower_95      = round(as.numeric(if (!is.null(fc$lower)) fc$lower[, 2] else fc$mean)),
    Upper_95      = round(as.numeric(if (!is.null(fc$upper)) fc$upper[, 2] else fc$mean))
  )

  cat(sprintf("  %s (%s): best=%s  MAPE=%.2f%%  LjungBox p=%.3f\n",
              s, site_name, best,
              min(c(hw_m["MAPE"], ar_m["MAPE"], sn_m["MAPE"])),
              lb_p))
}

# --- Write outputs -----------------------------------------------
fc_out <- bind_rows(all_forecasts)
metrics <- bind_rows(metrics_rows)
diag    <- bind_rows(diag_rows)

write_csv(fc_out, OUT_CSV)

sink(OUT_METR)
cat("Nova Leisure Group — Demand Model Accuracy (28-day holdout)\n")
cat(sprintf("Generated: %s\n", Sys.time()))
cat(paste(rep("=", 70), collapse = ""), "\n\n")
print(metrics, row.names = FALSE)
cat("\n\nBest model per site & residual whiteness (Ljung-Box p > 0.05 = white noise)\n")
cat(paste(rep("-", 70), collapse = ""), "\n")
print(diag, row.names = FALSE)
cat("\n\nGroup-level accuracy (visitor-weighted)\n")
cat(paste(rep("-", 70), collapse = ""), "\n")
grp <- metrics %>%
  group_by(Model) %>%
  summarise(Avg_RMSE = mean(RMSE), Avg_MAE = mean(MAE), Avg_MAPE = mean(MAPE)) %>%
  arrange(Avg_MAPE)
print(grp, row.names = FALSE)
sink()

# --- Plot --------------------------------------------------------
pdf(OUT_PLOT, width = 11, height = 7)
for (s in sites) {
  hist <- df %>% filter(Site_Code == s) %>% arrange(Date)
  fcs  <- fc_out %>% filter(Site_Code == s)
  plot(hist$Date, hist$Visitors, type = "l", col = "grey40",
       xlim = range(c(hist$Date, fcs$Date)),
       ylim = range(c(hist$Visitors, fcs$Lower_95, fcs$Upper_95), na.rm = TRUE),
       xlab = "Date", ylab = "Visitors",
       main = paste0(unique(hist$Site_Name), " — ", unique(fcs$Model), " forecast (90d)"))
  polygon(c(fcs$Date, rev(fcs$Date)),
          c(fcs$Lower_95, rev(fcs$Upper_95)),
          col = rgb(0.2, 0.4, 0.8, 0.15), border = NA)
  polygon(c(fcs$Date, rev(fcs$Date)),
          c(fcs$Lower_80, rev(fcs$Upper_80)),
          col = rgb(0.2, 0.4, 0.8, 0.25), border = NA)
  lines(fcs$Date, fcs$Forecast, col = "navy", lwd = 2)
  abline(v = max(hist$Date), col = "red", lty = 2)
  legend("topleft",
         legend = c("Historical", "Forecast", "80% PI", "95% PI"),
         col = c("grey40", "navy", rgb(0.2, 0.4, 0.8, 0.4), rgb(0.2, 0.4, 0.8, 0.2)),
         lwd = c(1, 2, 8, 8), bty = "n", cex = 0.8)
}
dev.off()

cat("\n--- WRITE COMPLETE ---\n")
cat(sprintf("Forecast CSV : %s  (%d rows)\n", OUT_CSV, nrow(fc_out)))
cat(sprintf("Metrics file : %s\n", OUT_METR))
cat(sprintf("Plot PDF     : %s\n", OUT_PLOT))

File: Predictive_Demand_Model/demand_forecast.py
"""
Nova Leisure Group — Predictive Demand Model (Python equivalent)
Mirrors demand_forecast.R for users without R installed.

Usage:  python3 demand_forecast.py
Outputs in Predictive_Demand_Model/:
  - forecast_output.csv
  - model_metrics.txt
  - forecast_plot.pdf
"""
import os, sys, warnings
import pandas as pd
import numpy as np
from datetime import timedelta

warnings.filterwarnings("ignore")

ROOT     = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
INPUT    = os.path.join(ROOT, "Daily_Site_Footfall.csv")
OUTDIR   = os.path.join(ROOT, "Predictive_Demand_Model")
os.makedirs(OUTDIR, exist_ok=True)
OUT_CSV  = os.path.join(OUTDIR, "forecast_output.csv")
OUT_METR = os.path.join(OUTDIR, "model_metrics.txt")
OUT_PLOT = os.path.join(OUTDIR, "forecast_plot.pdf")

HORIZON  = 90
HOLDOUT  = 28
np.random.seed(42)

from statsmodels.tsa.holtwinters import ExponentialSmoothing
from statsmodels.tsa.arima.model import ARIMA
from statsmodels.stats.diagnostic import acorr_ljungbox
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages

df = pd.read_csv(INPUT, parse_dates=["Date"]).sort_values(["Site_Code","Date"])
print(f"Loaded {len(df):,} daily rows across {df['Site_Code'].nunique()} sites.")

def metrics(y, yhat):
    err = np.array(y) - np.array(yhat)
    rmse = float(np.sqrt(np.mean(err**2)))
    mae  = float(np.mean(np.abs(err)))
    mape = float(np.mean(np.abs(err / np.where(np.array(y)==0, np.nan, np.array(y)))) * 100)
    return {"RMSE": rmse, "MAE": mae, "MAPE": mape}

def fit_arima(train):
    """Simple seasonal ARIMA(1,1,1)(1,1,1,7) — robust default that auto-arima typically selects."""
    try:
        m = ARIMA(train, order=(1,1,1), seasonal_order=(1,1,1,7),
                  enforce_stationarity=False, enforce_invertibility=False).fit()
        return m
    except Exception:
        m = ARIMA(train, order=(1,1,1)).fit()
        return m

def fit_hw(train):
    return ExponentialSmoothing(train, trend="add", seasonal="add", seasonal_periods=7,
                                initialization_method="estimated").fit(optimized=True)

def snaive(train, h, period=7):
    last_period = list(train[-period:])
    return [last_period[i % period] for i in range(h)]

all_fc = []
metrics_rows = []
diag_rows = []
plots = []

sites = df[["Site_Code","Site_Name","Region"]].drop_duplicates().to_records(index=False)

for site_code, site_name, region in sites:
    sdf = df[df["Site_Code"] == site_code].copy().reset_index(drop=True)
    y = sdf["Visitors"].astype(float).values
    n = len(y)
    train, test = y[:-HOLDOUT], y[-HOLDOUT:]

    # Holt-Winters
    try:
        hw_fit  = fit_hw(train)
        hw_pred = hw_fit.forecast(HOLDOUT)
    except Exception as e:
        hw_pred = np.repeat(np.mean(train), HOLDOUT)
        hw_fit  = None

    # ARIMA
    try:
        ar_fit  = fit_arima(train)
        ar_pred = ar_fit.forecast(HOLDOUT)
    except Exception:
        ar_pred = np.repeat(np.mean(train), HOLDOUT)
        ar_fit  = None

    # Seasonal Naive
    sn_pred = snaive(train, HOLDOUT)

    hw_m = metrics(test, hw_pred)
    ar_m = metrics(test, ar_pred)
    sn_m = metrics(test, sn_pred)

    for model, m in [("HoltWinters", hw_m), ("ARIMA", ar_m), ("SeasonalNaive", sn_m)]:
        metrics_rows.append({"Site_Code":site_code, "Site_Name":site_name,
                             "Model":model, **m})

    cands = {"HoltWinters": hw_m["MAPE"], "ARIMA": ar_m["MAPE"], "SeasonalNaive": sn_m["MAPE"]}
    best  = min(cands, key=cands.get)
    best_mape = cands[best]

    # Ljung-Box
    lb_p = None
    if best == "HoltWinters" and hw_fit is not None:
        resid = hw_fit.resid
    elif best == "ARIMA" and ar_fit is not None:
        resid = ar_fit.resid
    else:
        resid = None
    if resid is not None:
        try:
            lb = acorr_ljungbox(resid.dropna() if hasattr(resid,'dropna') else resid, lags=[14], return_df=True)
            lb_p = float(lb["lb_pvalue"].iloc[0])
        except Exception:
            lb_p = None

    diag_rows.append({"Site_Code":site_code, "Best_Model":best,
                      "LjungBox_p_lag14": lb_p})

    # Refit on full data → forecast HORIZON ahead with intervals (HW or ARIMA)
    if best == "ARIMA":
        m = fit_arima(y)
        fc_res = m.get_forecast(HORIZON)
        mean   = np.asarray(fc_res.predicted_mean)
        ci80   = np.asarray(fc_res.conf_int(alpha=0.20))
        ci95   = np.asarray(fc_res.conf_int(alpha=0.05))
        lo80, hi80 = ci80[:,0], ci80[:,1]
        lo95, hi95 = ci95[:,0], ci95[:,1]
    elif best == "HoltWinters":
        m = fit_hw(y)
        mean = np.array(m.forecast(HORIZON))
        # HW doesn't ship native PIs in statsmodels — bootstrap residuals
        res = m.resid.dropna().values if hasattr(m.resid,'dropna') else np.asarray(m.resid)
        sd  = np.std(res)
        lo80 = mean - 1.282 * sd; hi80 = mean + 1.282 * sd
        lo95 = mean - 1.960 * sd; hi95 = mean + 1.960 * sd
    else:
        mean = np.array(snaive(y, HORIZON))
        res  = np.array(y[-HOLDOUT:]) - np.array(snaive(y[:-HOLDOUT], HOLDOUT))
        sd   = np.std(res)
        lo80 = mean - 1.282 * sd; hi80 = mean + 1.282 * sd
        lo95 = mean - 1.960 * sd; hi95 = mean + 1.960 * sd

    last_date = sdf["Date"].iloc[-1]
    fc_dates  = [last_date + timedelta(days=i+1) for i in range(HORIZON)]

    for i in range(HORIZON):
        all_fc.append({
            "Date": fc_dates[i].date().isoformat(),
            "Site_Code": site_code,
            "Site_Name": site_name,
            "Region":    region,
            "Model":     best,
            "Forecast":  int(round(max(0, mean[i]))),
            "Lower_80":  int(round(max(0, lo80[i]))),
            "Upper_80":  int(round(max(0, hi80[i]))),
            "Lower_95":  int(round(max(0, lo95[i]))),
            "Upper_95":  int(round(max(0, hi95[i]))),
        })

    plots.append((site_code, site_name, sdf["Date"].values, y, fc_dates, mean, lo95, hi95, best))
    print(f"  {site_code} ({site_name}): best={best:13s}  MAPE={best_mape:5.2f}%  "
          f"LjungBox p={lb_p if lb_p is not None else float('nan'):.3f}")

# Save CSV
fc_df = pd.DataFrame(all_fc)
fc_df.to_csv(OUT_CSV, index=False)

# Save metrics text
mdf = pd.DataFrame(metrics_rows)
ddf = pd.DataFrame(diag_rows)
grp = mdf.groupby("Model")[["RMSE","MAE","MAPE"]].mean().sort_values("MAPE")

with open(OUT_METR, "w") as f:
    f.write("Nova Leisure Group — Demand Model Accuracy (28-day holdout)\n")
    f.write("=" * 70 + "\n\n")
    f.write("Per-site accuracy:\n")
    f.write(mdf.round(2).to_string(index=False))
    f.write("\n\nBest model per site & Ljung-Box residual whiteness (p > 0.05 = white noise):\n")
    f.write(ddf.round(3).to_string(index=False))
    f.write("\n\nGroup-level average accuracy (ranked):\n")
    f.write(grp.round(2).to_string())
    f.write("\n")

# Plot
with PdfPages(OUT_PLOT) as pdf:
    for site_code, site_name, hist_dates, hist_y, fc_dates, mean, lo95, hi95, best in plots:
        fig, ax = plt.subplots(figsize=(11, 6))
        ax.plot(hist_dates, hist_y, color="grey", linewidth=0.8, label="Historical")
        ax.fill_between(fc_dates, lo95, hi95, alpha=0.18, color="navy", label="95% PI")
        ax.plot(fc_dates, mean, color="navy", linewidth=2, label="Forecast")
        ax.axvline(hist_dates[-1], color="red", linestyle="--", linewidth=0.8)
        ax.set_title(f"{site_name} — {best} forecast (90d)")
        ax.set_xlabel("Date"); ax.set_ylabel("Visitors")
        ax.legend(loc="upper left", frameon=False)
        ax.grid(alpha=0.2)
        plt.tight_layout()
        pdf.savefig(fig); plt.close(fig)

print(f"\nWrote: {OUT_CSV}  ({len(fc_df):,} rows)")
print(f"Wrote: {OUT_METR}")
print(f"Wrote: {OUT_PLOT}")
